nvme_tcp.go

```diff
@@ -0,0 +1,658 @@
+	defer tracer.TraceFuncCall(ctx, "NVMeTCPConnector.ConnectVolume")()
+	if err := c.limiter.Acquire(ctx, 1); err != nil {
+		return Device{}, errors.New("too many parallel operations. try later")
+	}
+	defer c.limiter.Release(1)
+	addDefaultNVMeTCPPortToVolumeInfoPortals(&info)
+
+	if err := c.validateNVMeTCPVolumeInfo(ctx, info); err != nil {
+		return Device{}, err
+	}
+
+	ret, err, _ := c.singleCall.Do(
+		singleCallKeyForNVMeTCPTargets(info),
+		func() (interface{}, error) { return c.checkNVMeTCPSessions(ctx, info) })
+	if err != nil {
+		return Device{}, err
+	}
+	sessions := ret.([]gonvme.NVMESession)
+
+	ret, _, _ = c.singleCall.Do(
+		"IsDaemonRunning",
+		func() (interface{}, error) { return c.multipath.IsDaemonRunning(ctx), nil })
+	multipathIsEnabled := ret.(bool)
+
+	var d Device
+
+	if multipathIsEnabled {
+		logger.Info(ctx, "start multipath device connection")
+		d, err = c.connectMultipathDevice(ctx, sessions, info) //inprogress
+	} else {
+		logger.Info(ctx, "start single device connection")
+		//d, err = c.connectSingleDevice(ctx, sessions, info)
+	}
+
+	if err == nil {
+		if c.scsi.CheckDeviceIsValid(ctx, path.Join("/dev/", d.Name)) {
+			return d, nil
+		}
+		msg := fmt.Sprintf("device %s found but failed to read data from it", d.Name)
+		logger.Error(ctx, msg)
+		err = errors.New(msg)
+	}
+	logger.Error(ctx, "failed to connect volume, try to cleanup: %s", err.Error())
+	_ = c.cleanConnection(ctx, true, info)
+	return Device{}, err
+}
+
+func (c *NVMeTCPConnector) DisconnectVolume(ctx context.Context, info NVMeTCPVolumeInfo) error {
+	defer tracer.TraceFuncCall(ctx, "NVMeTCPConnector.DisconnectVolume")()
+	if err := c.limiter.Acquire(ctx, 1); err != nil {
+		return errors.New("too many parallel operations. try later")
+	}
+	defer c.limiter.Release(1)
+	addDefaultNVMeTCPPortToVolumeInfoPortals(&info)
+	return c.cleanConnection(ctx, false, info)
+}
+
+func (c *NVMeTCPConnector) DisconnectVolumeByDeviceName(ctx context.Context, name string) error {
+	defer tracer.TraceFuncCall(ctx, "NVMeTCPConnector.DisconnectVolumeByDeviceName")()
+	if err := c.limiter.Acquire(ctx, 1); err != nil {
+		return errors.New("too many parallel operations. try later")
+	}
+	defer c.limiter.Release(1)
+	return c.baseConnector.disconnectNVMEDevicesByDeviceName(ctx, name)
+}
+
+func (c *NVMeTCPConnector) GetInitiatorName(ctx context.Context) ([]string, error) {
+	defer tracer.TraceFuncCall(ctx, "NVMeTCPConnector.GetInitiatorName")()
+	logger.Info(ctx, "get initiator name")
+	data, err := c.nvmeTCPLib.GetInitiators("")
+	if err != nil {
+		logger.Error(ctx, "failed to read initiator name: %s", err.Error())
+	}
+	logger.Info(ctx, "initiator name is: %s", data)
+	return data, nil
+}
+
+func addDefaultNVMeTCPPortToVolumeInfoPortals(info *NVMeTCPVolumeInfo) {
+	for i, t := range info.Targets {
+		if !strings.Contains(t.Portal, ":") {
+			info.Targets[i].Portal += ":4420"
```

Comment:
I think maintaining this in a local constant inside the package would be a good idea even though this is a fixed port number, Because this port may be re-used in the future and few may provide `:4420` or just `4420`

Reply:
done. Moved to a constant